### ISSUE TITLE
Tracker 0.2.9 does not work

### ISSUE THREAD

**Issue by devmotion** [TuringLang/DistributionsAD.jl#107]
Tests of DistributionsAD (master branch) fail with Tracker 0.2.9 but pass with Tracker 0.2.8 (just tested locally).

**Comment by devmotion** [comment]
@willtebbutt @DhairyaLGandhi Any idea what might have code this breakage? Tracker-related code (a lot of type piracy :cry:) is contained in https://github.com/TuringLang/DistributionsAD.jl/blob/master/src/tracker.jl

**Comment by ChrisRackauckas** [comment]
DiffEqSensitivity doesn't do type-piracy and that one fails, so that's a sign.

**Comment by willtebbutt** [comment]
@devmotion this is unfortunate. Could you point us towards the tests in question?

**Comment by devmotion** [comment]
You can, e.g., compare https://github.com/TuringLang/DistributionsAD.jl/pull/106/checks?check_run_id=971611652 (Tracker 0.2.9) with https://github.com/TuringLang/DistributionsAD.jl/runs/929313813 (Tracker 0.2.8). The test summary of the new version is
```julia
Test Summary:                 | Pass  Fail  Error  Broken  Total
distributions                 | 1824  1798     78     310   4010
  Univariate distributions    |  207   257             44    508
  Multivariate distributions  |  643   437     30      89   1199
  Matrixvariate distributions |  974  1104     48     177   2303
```
This illustrates that it affects a large range of univariate, multivariate, and matrixvariate distributions.

The test file is https://github.com/TuringLang/DistributionsAD.jl/blob/master/test/ad/distributions.jl (it is used for testing the other AD backends such as Zygote, ForwardDiff, and ReverseDiff), and the first failing distributions seem to be https://github.com/TuringLang/DistributionsAD.jl/blob/master/test/ad/distributions.jl#L44-L45.

**Comment by DhairyaLGandhi** [comment]
I've opened FluxML/Tracker.jl#77 to debug if it's a change in behaviour we need to worry about. Does DistributionsAD make use of internal functions from Tracker? There was a change in the fields for `TrackedArray` but hopefully that's not what would cause this.

**Comment by ChrisRackauckas** [comment]
The DiffEqSensitivity test doesn't do any type piracy nor does it depend on DistributionsAD and it also fails, so that points to something internal to Tracker.

**Comment by devmotion** [comment]
> There was a change in the fields for TrackedArray but hopefully that's not what would cause this.

That caused some errors in Bijectors, but I've already fixed them: https://github.com/TuringLang/Bijectors.jl/pull/126/commits/33048d54c5581d4789a16dc3cbb2fca9a6570511

**Comment by devmotion** [comment]
> Does DistributionsAD make use of internal functions from Tracker?

A quick search didn't reveal anything yesterday.

**Comment by ChrisRackauckas** [comment]
For reference:

```julia
using DiffEqSensitivity, OrdinaryDiffEq, Tracker, ForwardDiff, FiniteDiff

function fb(du,u,p,t)
  du[1] = dx = p[1]*u[1] - p[2]*u[1]*u[2]*t
  du[2] = dy = -p[3]*u[2] + p[4]*u[1]*u[2]
end

p = [1.5,1.0,3.0,1.0]; u0 = [1.0;1.0]
prob = ODEProblem(fb,u0,(0.0,10.0),p)

function G(p)
  tmp_prob = remake(prob,u0=convert.(eltype(p),prob.u0),p=p)
  sol = solve(tmp_prob,Tsit5(),abstol=1e-14,reltol=1e-14,sensealg=DiffEqBase.SensitivityADPassThrough(),saveat=t)
  A = Array(sol)
  sum(((2 .- A).^2)./2)
end
G([1.5,1.0,3.0,1.0])
res2 = ForwardDiff.gradient(G,[1.5,1.0,3.0,1.0])
res3 = FiniteDiff.finite_difference_gradient(G,[1.5,1.0,3.0,1.0])
res4 = Tracker.gradient(G,[1.5,1.0,3.0,1.0])[1]
```

Gives

```julia
4-element Array{Float64,1}:
   228.68445307683288
  -237.74392662478962
   394.6268717258131
 -1223.4191564606485

4-element Array{Float64,1}:
   228.68445287934475
  -237.74392679565383
   394.6268717123165
 -1223.4191569287455

Tracked 4-element Array{Float64,1}:
  46.827325320171816
 -67.37120714519541
   4.689749988111205
 -12.586582069291765
```

One is not like the others.

**Comment by devmotion** [comment]
I just tried https://github.com/FluxML/Tracker.jl/pull/77 and it seems to fix the issues (https://github.com/FluxML/Tracker.jl/issues/76#issuecomment-672648467). Have you tried the MWE with the PR as well?

**Comment by DhairyaLGandhi** [comment]
Right, so FluxML/Tracker.jl#72 would need some work, we can revert it, fix downstream packages and open up a new PR for iterative backprop.

**Comment by willtebbutt** [comment]
Oh I see -- we had multiple PRs merged in the same patch release. I had assumed that it was something to do with me merging https://github.com/FluxML/Tracker.jl/pull/1, but I guess not.

**Comment by DhairyaLGandhi** [comment]
The new patch 0.2.10 should be on General soon now

**Comment by devmotion** [comment]
Fixed by Tracker 0.2.10.
